ledger/internal/evalprefetcher_alignment_test.go

```diff
@@ -0,0 +1,1295 @@
+
+	ch := prefetchAccounts(
+		context.Background(), l, 1,
+		[][]transactions.SignedTxnWithAD{group},
+		feeSink(), config.Consensus[proto])
+	loaded, ok := <-ch
+	require.True(t, ok)
+
+	require.NoError(t, loaded.err)
+	require.Equal(t, group, loaded.txnGroup)
+
+	_, ok = <-ch
+	require.False(t, ok)
+
+	accounts := parseLoadedAccountDataEntries(loaded.accounts)
+	apps, assets, creators := parseLoadedResourcesEntries(loaded.resources)
+	return accounts, apps, assets, creators
+}
+
+func runEval(t *testing.T, l *prefetcherAlignmentTestLedger, txn transactions.Transaction) {
+	genesisBlockHeader, err := l.BlockHdr(0)
+	require.NoError(t, err)
+	block := bookkeeping.MakeBlock(genesisBlockHeader)
+
+	eval, err := StartEvaluator(l, block.BlockHeader, EvaluatorOptions{})
+	require.NoError(t, err)
+
+	err = eval.TransactionGroup(makeGroupFromTxn(txn))
+	require.NoError(t, err)
+}
+
+func TestEvaluatorPrefetcherAlignmentPayment(t *testing.T) {
+	partitiontest.PartitionTest(t)
+
+	l := &prefetcherAlignmentTestLedger{
+		balances: map[basics.Address]ledgercore.AccountData{
+			rewardsPool(): {
+				AccountBaseData: ledgercore.AccountBaseData{
+					MicroAlgos: basics.MicroAlgos{Raw: 1234567890},
+				},
+			},
+			acctAddr(1): {
+				AccountBaseData: ledgercore.AccountBaseData{
+					MicroAlgos: basics.MicroAlgos{Raw: 1000001},
+				},
+			},
+			acctAddr(2): {
+				AccountBaseData: ledgercore.AccountBaseData{
+					MicroAlgos: basics.MicroAlgos{Raw: 1000002},
+				},
+			},
+			acctAddr(3): {
+				AccountBaseData: ledgercore.AccountBaseData{
+					MicroAlgos: basics.MicroAlgos{Raw: 1000003},
+				},
+			},
+		},
+	}
+
+	txn := transactions.Transaction{
+		Type: protocol.PaymentTx,
+		Header: transactions.Header{
+			Sender:      acctAddr(1),
+			GenesisHash: genesisHash(),
+		},
+		PaymentTxnFields: transactions.PaymentTxnFields{
+			Receiver:         acctAddr(2),
+			CloseRemainderTo: acctAddr(3),
+		},
+	}
+
+	accounts, apps, assets, creators := prefetch(t, l, txn)
+	runEval(t, l, txn)
+
+	accounts[rewardsPool()] = struct{}{}
+	require.Equal(t, l.requestedBalances, accounts)
+	require.Equal(t, l.requestedApps, apps)
+	require.Equal(t, l.requestedAssets, assets)
+	require.Equal(t, l.requestedCreators, creators)
+}
+
+func TestEvaluatorPrefetcherAlignmentCreateAsset(t *testing.T) {
+	partitiontest.PartitionTest(t)
+
+	l := &prefetcherAlignmentTestLedger{
+		balances: map[basics.Address]ledgercore.AccountData{
+			rewardsPool(): {
+				AccountBaseData: ledgercore.AccountBaseData{
+					MicroAlgos: basics.MicroAlgos{Raw: 1234567890},
+				},
+			},
+			acctAddr(1): {
+				AccountBaseData: ledgercore.AccountBaseData{
+					MicroAlgos: basics.MicroAlgos{Raw: 1000001},
+				},
+			},
+		},
+	}
+
+	txn := transactions.Transaction{
+		Type: protocol.AssetConfigTx,
+		Header: transactions.Header{
+			Sender:      acctAddr(1),
+			GenesisHash: genesisHash(),
+		},
+	}
+
+	accounts, apps, assets, creators := prefetch(t, l, txn)
+	runEval(t, l, txn)
```

Comment:
shouldn't `runEval` clear out the previously loaded `requestedBalances` ? otherwise, it wouldn't be able to properly compare the groups.

Reply:
Each test has its own `l`.

Reply:
yes, of course. But when you call the `runEval`, you should have a clean slate. Otherwise, you're testing commulative sum rather than distinct groups.
( don't forget that the prefetcher is calling the ledger to perform the loading of the resources.. which fills up the `requestedBalances` fields.. )